sticker/src/tagger/mod.rs

```diff
@@ -0,0 +1,161 @@
+use std::borrow::{Borrow, BorrowMut};
+use std::collections::HashMap;
+use std::convert::TryInto;
+
+use failure::Fallible;
+use ndarray::{Array1, ArrayD, Axis};
+use sticker_encoders::{EncodingProb, SentenceDecoder};
+use tch::Device;
+
+use crate::encoders::Encoders;
+use crate::input::vectorizer::WordPieceVectorizer;
+use crate::input::SentenceWithPieces;
+use crate::model::BertModel;
+use crate::tensor::{NoLabels, TensorBuilder, Tensors};
+use crate::util::seq_len_to_mask;
+
+/// A sequence tagger.
+pub struct Tagger {
+    device: Device,
+    encoders: Encoders,
+    model: BertModel,
+    vectorizer: WordPieceVectorizer,
+}
+
+impl Tagger {
+    /// Construct a new tagger.
+    pub fn new(
+        device: Device,
+        model: BertModel,
+        encoders: Encoders,
+        vectorizer: WordPieceVectorizer,
+    ) -> Self {
+        Tagger {
+            device,
+            model,
+            encoders,
+            vectorizer,
+        }
+    }
+
+    /// Tag sentences.
+    pub fn tag_sentences(
+        &self,
+        sentences: &mut [impl BorrowMut<SentenceWithPieces<String>>],
+    ) -> Fallible<()> {
+        let top_k_numeric = self.top_k_numeric_(sentences)?;
+
+        for (top_k, sentence) in top_k_numeric.into_iter().zip(sentences.iter_mut()) {
+            let sentence = sentence.borrow_mut();
+
+            for encoder in &*self.encoders {
+                let encoder_top_k = &top_k[encoder.name()];
+                encoder
+                    .encoder()
+                    .decode(&encoder_top_k, &mut sentence.sentence)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Construct the tensor representations of a batch of sentences.
+    fn prepare_batch(
+        &self,
+        sentences: &[impl Borrow<SentenceWithPieces<String>>],
+    ) -> Fallible<Tensors> {
+        let max_seq_len = sentences
+            .iter()
+            .map(|sentence| sentence.borrow().pieces.len())
+            .max()
+            .unwrap_or(0);
+
+        let mut builder: TensorBuilder<NoLabels> = TensorBuilder::new(
+            sentences.len(),
+            max_seq_len,
+            self.encoders.iter().map(|encoder| encoder.name()),
+        );
+
+        for sentence in sentences {
+            let sentence = sentence.borrow();
+            let input = self.vectorizer.vectorize(&sentence.pieces);
+            let mut token_mask = Array1::zeros((input.len(),));
+            for token_idx in &sentence.token_offsets {
+                token_mask[*token_idx] = 1;
+            }
```

Comment:
More for my understanding: What effect does this loop have if `token_mask` is never used afterwards?

Reply:
Good catch! It has no effect. However, it turns that the token mask is also needed for label-less inputs when doing model distillation (to mask word pieces for which we don't care about the divergence of the teacher model). So, I'd like to keep this around and do another PR soon that moves token masks from `LabelsTensor` to `TensorBuilder`.